compaction/compaction.cc

```diff
@@ -746,6 +746,45 @@ class compaction {
         return consumer(make_compacting_reader(setup_sstable_reader(), compaction_time, max_purgeable_func(), gc_state));
     }
 
+    // when compacting a run of mutation fragments, the producer reads the
+    // mutations from sstables (or yet another reader). and the consumer
+    // processes the mutations, by either compacting them, or segregates them
+    // into chunks. each chunk of the fragments stream can be either piped to
+    // a sstable writer or yet another consumer. the consumers and writers can
+    // be created on demand.
+    //
+    //   def sstable_reader(sstables):
+    //       for sstable in sstables:
+    //           for frag in read(sstable):
+    //               yield frag
```

Comment:
All sstables are joined up by a combined reader, which combines their output, such that fragments with the same position are merged.

Reply:
This is similar to python's `zip()`, except the result is not a tuple, the tuple is squashed into a single element.

Reply:
how about:

```py
def sstable_reader(sstables):
    all_frags = chain(read(sstable) for sstable in sstables)
    all_frags = sorted(all_frags, key=frag.position)
    for frags in groupby(all_frags, key=frag.position):
        yield merge(frags)
```
i am trying to get a balance between the readability and accuracy.

Reply:
I think this reflects the reality better:
```python
for frag_tuple in zip(*sstable_readers):
   frag0 = frag_tuple[0]
   for frag in frag_tuple[1:]:
      frag0.apply(frag)
   yield frag0
```

Reply:
i am confused. assuming that `read(sstable)` is an `Iterable[Fragment]`, and different instances of `sstable` may contains different number of fragments, so what do we expect from zipping them up? for instance, if we are compacting 3 sstables:

- sstable_1: [0, 1, 2, 4]
- sstable_2: [5, 6]
- sstable_3: [7, 9, 12]

if we zip them like `zip(sstable_1, sstable_2, sstable_3)`, we'd get `[(0, 5, 7), (1, 6, 9)]`. applying your code to this dataset, we will have:
```py
for frags in [(0, 5, 7), (1, 6, 9)]:
    # applied = apply_onto(5, 7)
    # applied = apply_onto(applied, 7)
    # ...
    yield functools.reduce(frag.apply_onto, frags)
```

this, IMHO, does not make sense.

if the sstables are disjointed, agree that there is no need to sort them *after* reading the sstables, we just need to sort by the starting position of each sstable.

Reply:
It's a k-way merge. https://en.wikipedia.org/wiki/K-way_merge_algorithm

if sstables are disjoint, we're able to consume one sstable at a time 

if they're not, we consume data from sstables that contain the current token.

A good exercise is to understand how incremental_reader_selector interacts with combined_reader.

Reply:
Like @raphaelsc noted above, the combined reader does a K-way merge. On each call to its generator method `operator()` it returns a tuple, with size: `[1, N]`, where `N` is the number of readers. The tuple will have an element for each reader, which had a fragment for the currently processed position in the stream.

Reply:
thank you, Botond and Raphael. i think that's exactly what i wanted to explain using the sort + groupby. but i agree, it's far from perfect. i will drop the `sstable_reader()` from the comment. as it seems that it'd be hard to achieve the balance of accuracy and simplicity.

Reply:
I think you can consider the sstables as a unified (merged) stream of fragments, for the purposes of this documentation. The exact way the sstables are combined is an unnecessary detail for compaction.